src/mr/master.go

```diff
@@ -1,29 +1,115 @@
 package mr
 
-import "log"
-import "net"
-import "os"
-import "net/rpc"
-import "net/http"
+import (
+	"log"
+	"net"
+	"net/http"
+	"net/rpc"
+	"os"
+	"sync"
+	"time"
+)
+
+type taskPool struct {
+	pending map[int]interface{}
+	issued  map[int]interface{}
+	mutex   sync.Mutex
+}
+
+func makeTaskPool() taskPool {
+	return taskPool{pending: make(map[int]interface{}), issued: make(map[int]interface{})}
+}
+
+func (tp *taskPool) insert(id int, task interface{}) {
```

Comment:
Use cmap that does exactly this https://pkg.go.dev/github.com/orcaman/concurrent-map